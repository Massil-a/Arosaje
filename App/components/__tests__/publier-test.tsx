import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react-native';
import PublierScreen from '../../app/(tabs)/publier';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { NavigationContainer } from '@react-navigation/native';
jest.mock('@react-native-async-storage/async-storage', () => ({
    setItem: jest.fn(),
    getItem: jest.fn(),
    removeItem: jest.fn(),
    clear: jest.fn(),
}));
jest.mock('@react-native-async-storage/async-storage');

describe('PublierScreen', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const { getByText } = render(
      <NavigationContainer>
        <PublierScreen />
      </NavigationContainer>
    );
    expect(getByText('Valider')).toBeTruthy(); // Vérifie que le bouton "Valider" est présent
  });

  it('opens modal on "Valider" press', () => {
    const { getByText } = render(
      <NavigationContainer>
        <PublierScreen />
      </NavigationContainer>
    );

    fireEvent.press(getByText('Valider'));
  });

  it('handles sending post correctly', async () => {
    const { getByText } = render(
      <NavigationContainer>
        <PublierScreen />
      </NavigationContainer>
    );

    // Simuler la saisie des données
    await AsyncStorage.setItem('userToken', 'mockedToken');
    fireEvent.changeText(getByText('Titre : '), 'Titre de test');
    fireEvent.changeText(getByText('Description : '), 'Description de test');
    fireEvent.changeText(getByText('Localisation : '), 'Localisation de test');
    fireEvent.changeText(getByText('Espèce : '), 'Espèce de test');
    fireEvent.changeText(getByText('Exigence d’entretien (optionel)'), 'Entretien de test');

    // Ouvrir la modal et simuler l'envoi
    fireEvent.press(getByText('Valider'));

  });

  it('shows error alert on missing user token', async () => {
    const { getByText } = render(
      <NavigationContainer>
        <PublierScreen />
      </NavigationContainer>
    );

    fireEvent.press(getByText('Valider'));
  }); // Correction ici : fermeture de la parenthèse manquante

  it('shows error alert on post send failure', async () => {
    const { getByText } = render(
      <NavigationContainer>
        <PublierScreen />
      </NavigationContainer>
    );

    await AsyncStorage.setItem('userToken', 'mockedToken');
    fireEvent.press(getByText('Valider'));

    // Simuler une réponse d'erreur du serveur
    global.fetch = jest.fn(() =>
      Promise.resolve({
        ok: false,
        status: 500, // Ajoutez le statut
        statusText: 'Internal Server Error', // Ajoutez le texte du statut
        headers: new Headers(), // Ajoutez les en-têtes
      } as Response) // Cast en tant que Response
    );

  });

  it('resets form on confirmation', async () => {
    const { getByText } = render(
      <NavigationContainer>
        <PublierScreen />
      </NavigationContainer>
    );

    await AsyncStorage.setItem('userToken', 'mockedToken');
    fireEvent.changeText(getByText('Titre : '), 'Titre de test');
    fireEvent.press(getByText('Recommencez'));



    // Simuler la saisie de données valides
    await AsyncStorage.setItem('userToken', 'mockedToken');
    fireEvent.changeText(getByText('Titre : '), 'Titre de test');
    fireEvent.changeText(getByText('Description : '), 'Description de test');
    fireEvent.changeText(getByText('Localisation : '), 'Localisation de test');
    fireEvent.changeText(getByText('Espèce : '), 'Espèce de test');

  });

  it('handles modal close correctly', async () => {
    const { getByText } = render(
      <NavigationContainer>
        <PublierScreen />
      </NavigationContainer>
    );

    fireEvent.press(getByText('Valider'));

  });
});
