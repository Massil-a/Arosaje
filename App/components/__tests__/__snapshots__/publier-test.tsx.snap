// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNSScreenStack
    style={
      {
        "flex": 1,
      }
    }
  >
    <RNSScreen
      collapsable={false}
      gestureResponseDistance={
        {
          "bottom": -1,
          "end": -1,
          "start": -1,
          "top": -1,
        }
      }
      nativeBackButtonDismissalEnabled={false}
      onAppear={[Function]}
      onDisappear={[Function]}
      onDismissed={[Function]}
      onGestureCancel={[Function]}
      onHeaderBackButtonClicked={[Function]}
      onNativeDismissCancelled={[Function]}
      onTransitionProgress={[Function]}
      onWillDisappear={[Function]}
      replaceAnimation="push"
      sheetAllowedDetents="large"
      sheetCornerRadius={-1}
      sheetExpandsWhenScrolledToEdge={true}
      sheetGrabberVisible={false}
      sheetLargestUndimmedDetent="all"
      stackPresentation="push"
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      swipeDirection="horizontal"
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        style={
          {
            "flex": 1,
            "flexDirection": "column-reverse",
          }
        }
      >
        <View
          collapsable={false}
          style={
            [
              {
                "flex": 1,
              },
              {
                "backgroundColor": "rgb(242, 242, 242)",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              {
                "backgroundColor": "#FFF",
                "flex": 1,
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "flex-start",
                  "marginTop": 30,
                }
              }
            >
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  [
                    {
                      "color": "#BDBDBD",
                      "padding": 10,
                    },
                    {
                      "width": "100%",
                    },
                  ]
                }
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                    }
                  }
                >
                  <Text
                    style={
                      {
                        "fontSize": 14,
                      }
                    }
                  >
                    Titre : 
                  </Text>
                  <Text
                    allowFontScaling={false}
                    selectable={false}
                    style={
                      [
                        {
                          "color": "black",
                          "fontSize": 18,
                        },
                        undefined,
                        {
                          "fontFamily": "ionicons",
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        },
                        {},
                      ]
                    }
                  >
                    
                  </Text>
                </View>
              </View>
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": "#ff2b24",
                      "fontSize": 24,
                    },
                    {
                      "paddingHorizontal": 10,
                    },
                    {
                      "fontFamily": "ionicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
              >
                
              </Text>
              <View
                style={
                  {
                    "alignSelf": "center",
                    "backgroundColor": "#E8E8E8",
                    "height": 1,
                    "marginVertical": 5,
                    "width": "95%",
                  }
                }
              />
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  [
                    {
                      "color": "#BDBDBD",
                      "padding": 10,
                    },
                    {
                      "width": "100%",
                    },
                  ]
                }
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                    }
                  }
                >
                  <Text
                    style={
                      {
                        "fontSize": 14,
                      }
                    }
                  >
                    Date(s) :  - 
                  </Text>
                  <Text
                    allowFontScaling={false}
                    selectable={false}
                    style={
                      [
                        {
                          "color": "black",
                          "fontSize": 18,
                        },
                        undefined,
                        {
                          "fontFamily": "ionicons",
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        },
                        {},
                      ]
                    }
                  >
                    
                  </Text>
                </View>
              </View>
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": "#ff2b24",
                      "fontSize": 24,
                    },
                    {
                      "paddingHorizontal": 10,
                    },
                    {
                      "fontFamily": "ionicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
              >
                
              </Text>
              <View
                style={
                  {
                    "alignSelf": "center",
                    "backgroundColor": "#E8E8E8",
                    "height": 1,
                    "marginVertical": 5,
                    "width": "95%",
                  }
                }
              />
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  [
                    {
                      "color": "#BDBDBD",
                      "padding": 10,
                    },
                    {
                      "width": "100%",
                    },
                  ]
                }
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                    }
                  }
                >
                  <Text
                    style={
                      {
                        "fontSize": 14,
                      }
                    }
                  >
                    Photo :  
                  </Text>
                  <Text
                    allowFontScaling={false}
                    selectable={false}
                    style={
                      [
                        {
                          "color": "black",
                          "fontSize": 18,
                        },
                        undefined,
                        {
                          "fontFamily": "ionicons",
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        },
                        {},
                      ]
                    }
                  >
                    
                  </Text>
                </View>
              </View>
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": "#ff2b24",
                      "fontSize": 24,
                    },
                    {
                      "paddingHorizontal": 10,
                    },
                    {
                      "fontFamily": "ionicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
              >
                
              </Text>
              <View
                style={
                  {
                    "alignSelf": "center",
                    "backgroundColor": "#E8E8E8",
                    "height": 1,
                    "marginVertical": 5,
                    "width": "95%",
                  }
                }
              />
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  [
                    {
                      "color": "#BDBDBD",
                      "padding": 10,
                    },
                    {
                      "width": "100%",
                    },
                  ]
                }
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                    }
                  }
                >
                  <Text
                    style={
                      {
                        "fontSize": 14,
                      }
                    }
                  >
                    Description : 
                  </Text>
                  <Text
                    allowFontScaling={false}
                    selectable={false}
                    style={
                      [
                        {
                          "color": "black",
                          "fontSize": 18,
                        },
                        undefined,
                        {
                          "fontFamily": "ionicons",
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        },
                        {},
                      ]
                    }
                  >
                    
                  </Text>
                </View>
              </View>
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": "#ff2b24",
                      "fontSize": 24,
                    },
                    {
                      "paddingHorizontal": 10,
                    },
                    {
                      "fontFamily": "ionicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
              >
                
              </Text>
              <View
                style={
                  {
                    "alignSelf": "center",
                    "backgroundColor": "#E8E8E8",
                    "height": 1,
                    "marginVertical": 5,
                    "width": "95%",
                  }
                }
              />
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  [
                    {
                      "color": "#BDBDBD",
                      "padding": 10,
                    },
                    {
                      "width": "100%",
                    },
                  ]
                }
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                    }
                  }
                >
                  <Text
                    style={
                      {
                        "fontSize": 14,
                      }
                    }
                  >
                    Localisation :  
                  </Text>
                  <Text
                    allowFontScaling={false}
                    selectable={false}
                    style={
                      [
                        {
                          "color": "black",
                          "fontSize": 18,
                        },
                        undefined,
                        {
                          "fontFamily": "ionicons",
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        },
                        {},
                      ]
                    }
                  >
                    
                  </Text>
                </View>
              </View>
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": "#ff2b24",
                      "fontSize": 24,
                    },
                    {
                      "paddingHorizontal": 10,
                    },
                    {
                      "fontFamily": "ionicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
              >
                
              </Text>
              <View
                style={
                  {
                    "alignSelf": "center",
                    "backgroundColor": "#E8E8E8",
                    "height": 1,
                    "marginVertical": 5,
                    "width": "95%",
                  }
                }
              />
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  [
                    {
                      "color": "#BDBDBD",
                      "padding": 10,
                    },
                    {
                      "width": "100%",
                    },
                  ]
                }
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                    }
                  }
                >
                  <Text
                    style={
                      {
                        "fontSize": 14,
                      }
                    }
                  >
                    Espèce : 
                  </Text>
                  <Text
                    allowFontScaling={false}
                    selectable={false}
                    style={
                      [
                        {
                          "color": "black",
                          "fontSize": 18,
                        },
                        undefined,
                        {
                          "fontFamily": "ionicons",
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        },
                        {},
                      ]
                    }
                  >
                    
                  </Text>
                </View>
              </View>
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": "#ff2b24",
                      "fontSize": 24,
                    },
                    {
                      "paddingHorizontal": 10,
                    },
                    {
                      "fontFamily": "ionicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
              >
                
              </Text>
              <View
                style={
                  {
                    "alignSelf": "center",
                    "backgroundColor": "#E8E8E8",
                    "height": 1,
                    "marginVertical": 5,
                    "width": "95%",
                  }
                }
              />
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  [
                    {
                      "color": "#BDBDBD",
                      "padding": 10,
                    },
                    {
                      "width": "100%",
                    },
                  ]
                }
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                    }
                  }
                >
                  <Text
                    style={
                      {
                        "fontSize": 14,
                      }
                    }
                  >
                    Exigence d’entretien (optionel)
                  </Text>
                  <Text
                    allowFontScaling={false}
                    selectable={false}
                    style={
                      [
                        {
                          "color": "black",
                          "fontSize": 18,
                        },
                        undefined,
                        {
                          "fontFamily": "ionicons",
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        },
                        {},
                      ]
                    }
                  >
                    
                  </Text>
                </View>
              </View>
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": "#828282",
                      "fontSize": 24,
                    },
                    {
                      "paddingHorizontal": 10,
                    },
                    {
                      "fontFamily": "ionicons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
              >
                
              </Text>
              <View
                style={
                  {
                    "alignSelf": "center",
                    "backgroundColor": "#E8E8E8",
                    "height": 1,
                    "marginVertical": 5,
                    "width": "95%",
                  }
                }
              />
            </View>
            <View
              style={
                {
                  "alignItems": "center",
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                }
              }
            >
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "marginBottom": 10,
                    "marginTop": 20,
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "red",
                      "fontSize": 16,
                      "marginTop": 20,
                    }
                  }
                >
                  Recommencez
                </Text>
              </View>
              <View
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "#E0E0E0",
                    "borderRadius": 25,
                    "flexDirection": "row",
                    "marginBottom": 20,
                    "overflow": "hidden",
                    "width": "90%",
                  }
                }
              >
                <View
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": undefined,
                      "disabled": true,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    [
                      {
                        "alignItems": "center",
                        "flex": 1,
                        "justifyContent": "center",
                        "paddingVertical": 20,
                      },
                      {
                        "backgroundColor": "#828282",
                      },
                    ]
                  }
                >
                  <Text
                    style={
                      {
                        "color": "#FFF",
                        "fontSize": 14,
                        "fontWeight": "bold",
                      }
                    }
                  >
                    Valider
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
      <RNSScreenStackHeaderConfig
        backButtonInCustomView={false}
        backTitleVisible={true}
        backgroundColor="#668F80"
        color="#fff"
        direction="ltr"
        disableBackButtonMenu={false}
        hidden={false}
        hideBackButton={false}
        largeTitleHideShadow={false}
        title="Publier"
        titleColor="#FFF"
        titleFontSize={24}
        titleFontWeight="bold"
        topInsetEnabled={false}
        translucent={false}
      />
    </RNSScreen>
  </RNSScreenStack>
</RNCSafeAreaProvider>
`;
